Allow versioned resources to have failed previous versions.

**Describe the problem/challenge you have**

When I use versioned job resources for data migrations, it would be useful to allow failed previous versions. In my development workflow these data migrations often fail and require developers to diagnose and deploy new versions. Ideally in this workflow:
- Developers don't need to manually modify (i.e. delete) k8s resources. Our deployments are all managed by a CI system, triggered when developers push code to git.
- Developers can keep the history of the failed previous versions.

**Describe the solution you'd like**

Allow versioned resources to have failed previous versions. Perhaps through annotation configuration.

**Anything else you would like to add:**

This is a followup from Slack:
https://kubernetes.slack.com/archives/CH8KCCKA5/p1647720204911829

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

I also noticed that using the `--diff-changes` flag when there is already a failed version present in the app results in a panic as shown below.
```bash
kapp -n apps deploy -a versioned-job-test -f app-succeed.yaml -c                                                                             
Target cluster 'https://192.168.64.81:8443' (nodes: minikube)

@@ create job/my-job-ver-3 (batch/v1) namespace: versioned-job-test @@
  ...
  5,  5       kapp.k14s.io/versioned: ""
      6 +     kapp.k14s.io/versioned-keep-original: ""
  6,  7     creationTimestamp: "2022-03-21T06:18:16Z"
  7,  8     generation: 1
  ...
 90, 91           - -c
 91     -         - exit 1
     92 +         - exit 0
 92, 93           image: alpine:3.15.0
 93, 94           name: my-job
@@ noop job/my-job-ver-2 (batch/v1) namespace: versioned-job-test @@
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1fe8e95]

goroutine 1 [running]:
github.com/k14s/kapp/pkg/kapp/diff.TextDiffView.String({0x0, {0xc000d4f5c0, 0x2, 0x2}, {0x2, 0x1, 0x1}})
	github.com/k14s/kapp/pkg/kapp/diff/text_diff_view.go:37 +0xd5
github.com/k14s/kapp/pkg/kapp/clusterapply.(*ChangeSetView).Print(0xc000b18568, {0x26beba0, 0xc000425680})
	github.com/k14s/kapp/pkg/kapp/clusterapply/change_set_view.go:37 +0x387
github.com/k14s/kapp/pkg/kapp/cmd/app.(*DeployOptions).calculateAndPresentChanges(_, {_, _, _}, {_, _, _}, {{0xc000a8f200, 0x1, 0x1}}, ...)
	github.com/k14s/kapp/pkg/kapp/cmd/app/deploy.go:423 +0xc85
github.com/k14s/kapp/pkg/kapp/cmd/app.(*DeployOptions).Run(0xc0003e82c0)
	github.com/k14s/kapp/pkg/kapp/cmd/app/deploy.go:149 +0x8a5
github.com/k14s/kapp/pkg/kapp/cmd/app.NewDeployCmd.func1(0xc0007a0960, {0x243e817, 0xa, 0x226b5a0})
	github.com/k14s/kapp/pkg/kapp/cmd/app/deploy.go:55 +0x1d
github.com/k14s/kapp/pkg/kapp/cmd.NewKappCmd.func1.1(0xc0001d1c80, {0xc0001b84d0, 0x0, 0x7})
	github.com/k14s/kapp/pkg/kapp/cmd/kapp.go:128 +0xce
github.com/cppforlife/cobrautil.WrapRunEForCmd.func1.1(0x11f65a5, {0xc0001b84d0, 0x0, 0x7})
	github.com/cppforlife/cobrautil@v0.0.0-20200514214827-bb86e6965d72/misc.go:45 +0x6d
github.com/cppforlife/cobrautil.WrapRunEForCmd.func1.1(0xc000880400, {0xc0001b84d0, 0x0, 0x7})
	github.com/cppforlife/cobrautil@v0.0.0-20200514214827-bb86e6965d72/misc.go:45 +0x6d
github.com/cppforlife/cobrautil.WrapRunEForCmd.func1.1(0xc0004bb680, {0xc0001b84d0, 0x0, 0x7})
	github.com/cppforlife/cobrautil@v0.0.0-20200514214827-bb86e6965d72/misc.go:45 +0x6d
github.com/spf13/cobra.(*Command).execute(0xc0004bb680, {0xc0001b8460, 0x7, 0x7})
	github.com/spf13/cobra@v1.1.3/command.go:852 +0x60e
github.com/spf13/cobra.(*Command).ExecuteC(0xc0004bac80)
	github.com/spf13/cobra@v1.1.3/command.go:960 +0x3ad
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.1.3/command.go:897
main.nonExitingMain()
	github.com/k14s/kapp/cmd/kapp/kapp.go:42 +0x153
main.main()
	github.com/k14s/kapp/cmd/kapp/kapp.go:21 +0x19
```

--- a/pkg/kapp/diff/change_precalculated.go
+++ b/pkg/kapp/diff/change_precalculated.go
@@ -20,7 +20,9 @@ type ChangePrecalculated struct {
 
 var _ Change = &ChangePrecalculated{}
 
-func NewChangePrecalculated(existingRes, newRes, appliedRes ctlres.Resource) *ChangePrecalculated {
+func NewChangePrecalculated(existingRes, newRes, appliedRes ctlres.Resource,
+	op ChangeOp, configurableTextDiff *ConfigurableTextDiff, opsDiff OpsDiff) *ChangePrecalculated {
+
 	if existingRes == nil && newRes == nil {
 		panic("Expected either existingRes or newRes be non-nil")
 	}
@@ -35,7 +37,8 @@ func NewChangePrecalculated(existingRes, newRes, appliedRes ctlres.Resource) *Ch
 		appliedRes = appliedRes.DeepCopy()
 	}
 
-	return &ChangePrecalculated{existingRes: existingRes, newRes: newRes, appliedRes: appliedRes}
+	return &ChangePrecalculated{existingRes: existingRes, newRes: newRes, appliedRes: appliedRes,
+		op: op, configurableTextDiff: configurableTextDiff, opsDiff: opsDiff}
 }
 
 func (d *ChangePrecalculated) NewOrExistingResource() ctlres.Resource {

--- a/pkg/kapp/diff/change_set_with_versioned_rs.go
+++ b/pkg/kapp/diff/change_set_with_versioned_rs.go
@@ -42,21 +42,21 @@ func (d ChangeSetWithVersionedRs) Calculate() ([]Change, error) {
 
 	// First try to calculate changes will update references on all resources
 	// (which includes versioned and non-versioned resources)
-	_, _, err := d.addChanges(newRs, existingRsGrouped)
+	_, _, err := d.addAndKeepChanges(newRs, existingRsGrouped)
 	if err != nil {
 		return nil, err
 	}
 
 	// Since there might have been circular dependencies;
 	// second try catches ones that werent changed during first run
-	addChanges, alreadyAdded, err := d.addChanges(newRs, existingRsGrouped)
+	addChanges, alreadyAdded, err := d.addAndKeepChanges(newRs, existingRsGrouped)
 	if err != nil {
 		return nil, err
 	}
 
 	allChanges = append(allChanges, addChanges...)
 
-	keepAndDeleteChanges, err := d.keepAndDeleteChanges(existingRsGrouped, alreadyAdded)
+	keepAndDeleteChanges, err := d.noopAndDeleteChanges(existingRsGrouped, alreadyAdded)
 	if err != nil {
 		return nil, err
 	}
@@ -113,7 +113,7 @@ func (d ChangeSetWithVersionedRs) assignNewNames(
 	}
 }
 
-func (d ChangeSetWithVersionedRs) addChanges(
+func (d ChangeSetWithVersionedRs) addAndKeepChanges(
 	newRs versionedResources, existingRsGrouped map[string][]ctlres.Resource) (
 	[]Change, map[string]ctlres.Resource, error) {
 
@@ -139,6 +139,7 @@ func (d ChangeSetWithVersionedRs) addChanges(
 			case ChangeOpKeep:
 				// Use latest copy of resource to update affected resources
 				usedRes = existingRes
+				changes = append(changes, d.newKeepChange(existingRes))
 			default:
 				panic(fmt.Sprintf("Unexpected change op %s", updateChange.Op()))
 			}
@@ -182,7 +183,7 @@ func (d ChangeSetWithVersionedRs) newAddChangeFromUpdateChange(
 	return addChange
 }
 
-func (d ChangeSetWithVersionedRs) keepAndDeleteChanges(
+func (d ChangeSetWithVersionedRs) noopAndDeleteChanges(
 	existingRsGrouped map[string][]ctlres.Resource,
 	alreadyAdded map[string]ctlres.Resource) ([]Change, error) {
 
@@ -212,9 +213,9 @@ func (d ChangeSetWithVersionedRs) keepAndDeleteChanges(
 			changes = append(changes, change)
 		}
 
-		// Create changes that "keep" resources
+		// Create changes that "noop" resources
 		for _, existingRes := range existingRs[len(existingRs)-numToKeep:] {
-			changes = append(changes, d.newKeepChange(existingRes))
+			changes = append(changes, d.newNoopChange(existingRes))
 		}
 	}
 
@@ -226,6 +227,14 @@ func (d ChangeSetWithVersionedRs) newKeepChange(existingRes ctlres.Resource) Cha
 	addChange := NewChangePrecalculated(existingRes, nil, nil)
 	// TODO private field access
 	addChange.op = ChangeOpKeep
+	return addChange
+}
+
+func (d ChangeSetWithVersionedRs) newNoopChange(existingRes ctlres.Resource) Change {
+	// Use update's diffs but create a change for new resource
+	addChange := NewChangePrecalculated(existingRes, nil, nil)
+	// TODO private field access
+	addChange.op = ChangeOpNoop
 	return addChange
 }
 

--- a/pkg/kapp/diff/change_set_with_versioned_rs.go
+++ b/pkg/kapp/diff/change_set_with_versioned_rs.go
@@ -175,12 +175,7 @@ func (d ChangeSetWithVersionedRs) newAddChangeFromUpdateChange(
 	newRes ctlres.Resource, updateChange Change) Change {
 
 	// Use update's diffs but create a change for new resource
-	addChange := NewChangePrecalculated(nil, newRes, newRes)
-	// TODO private field access
-	addChange.op = ChangeOpAdd
-	addChange.configurableTextDiff = updateChange.ConfigurableTextDiff()
-	addChange.opsDiff = updateChange.OpsDiff()
-	return addChange
+	return NewChangePrecalculated(nil, newRes, newRes, ChangeOpAdd, updateChange.ConfigurableTextDiff(), updateChange.OpsDiff())
 }
 
 func (d ChangeSetWithVersionedRs) noopAndDeleteChanges(
@@ -223,19 +218,11 @@ func (d ChangeSetWithVersionedRs) noopAndDeleteChanges(
 }
 
 func (d ChangeSetWithVersionedRs) newKeepChange(existingRes ctlres.Resource) Change {
-	// Use update's diffs but create a change for new resource
-	addChange := NewChangePrecalculated(existingRes, nil, nil)
-	// TODO private field access
-	addChange.op = ChangeOpKeep
-	return addChange
+	return NewChangePrecalculated(existingRes, nil, nil, ChangeOpKeep, NewConfigurableTextDiff(existingRes, nil, true), OpsDiff{})
 }
 
 func (d ChangeSetWithVersionedRs) newNoopChange(existingRes ctlres.Resource) Change {
-	// Use update's diffs but create a change for new resource
-	addChange := NewChangePrecalculated(existingRes, nil, nil)
-	// TODO private field access
-	addChange.op = ChangeOpNoop
-	return addChange
+	return NewChangePrecalculated(existingRes, nil, nil, ChangeOpNoop, nil, OpsDiff{})
 }
 
 func (ChangeSetWithVersionedRs) numOfResourcesToKeep(res ctlres.Resource) (int, error) {

--- a/test/e2e/annotations_test.go
+++ b/test/e2e/annotations_test.go
@@ -550,8 +550,8 @@ spec:
 			"Wait to: 1 reconcile, 0 delete, 0 noop", out)
 	})
 
-	logger.Section("deploy with no changes", func() {
-		out, err := kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name, "--json"},
+	logger.Section("deploy with no changes and use --diff-changes", func() {
+		out, err := kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name, "-c", "--json"},
 			RunOpts{IntoNs: true, AllowError: true, StdinReader: strings.NewReader(failYaml)})
 
 		require.Error(t, err, "Expected to receive an error")

--- a/test/e2e/annotations_test.go
+++ b/test/e2e/annotations_test.go
@@ -473,3 +473,126 @@ data:
 			"Wait to: 1 reconcile, 1 delete, 0 noop", kappOut)
 	})
 }
+
+func TestVersionedAnnotation_WithFailedPreviousVersions(t *testing.T) {
+	env := BuildEnv(t)
+	logger := Logger{}
+	kapp := Kapp{t, env.Namespace, env.KappBinaryPath, logger}
+
+	failYaml := `
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: my-job
+  annotations:
+    kapp.k14s.io/versioned: ""
+spec:
+  template:
+    metadata:
+      name: my-job
+    spec:
+      restartPolicy: Never
+      containers:
+        - name: my-job
+          image: alpine:3.15.0
+          command: [ "sh", "-c", "exit 1" ]
+  backoffLimit: 0
+`
+
+	successYaml := `
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: my-job
+  annotations:
+    kapp.k14s.io/versioned: ""
+spec:
+  template:
+    metadata:
+      name: my-job
+    spec:
+      restartPolicy: Never
+      containers:
+        - name: my-job
+          image: alpine:3.15.0
+          command: [ "sh", "-c", "exit 0" ]
+  backoffLimit: 0
+`
+
+	name := "test-ver-ann-with-failed-prev-ver"
+	cleanUp := func() {
+		kapp.Run([]string{"delete", "-a", name})
+	}
+
+	cleanUp()
+	defer cleanUp()
+
+	logger.Section("deploy ver-1 of job which is expected to fail", func() {
+		out, err := kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name, "--json"},
+			RunOpts{IntoNs: true, AllowError: true, StdinReader: strings.NewReader(failYaml)})
+
+		require.Error(t, err, "Expected to receive an error")
+
+		resp := uitest.JSONUIFromBytes(t, []byte(out))
+
+		expectedOutput := []map[string]string{{
+			"kind":            "Job",
+			"name":            "my-job-ver-1",
+			"namespace":       env.Namespace,
+			"op":              "create",
+			"op_strategy":     "",
+			"reconcile_info":  "",
+			"reconcile_state": "",
+			"wait_to":         "reconcile",
+		}}
+
+		validateChanges(t, resp.Tables, expectedOutput, "Op:      1 create, 0 delete, 0 update, 0 noop, 0 exists",
+			"Wait to: 1 reconcile, 0 delete, 0 noop", out)
+	})
+
+	logger.Section("deploy with no changes", func() {
+		out, err := kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name, "--json"},
+			RunOpts{IntoNs: true, AllowError: true, StdinReader: strings.NewReader(failYaml)})
+
+		require.Error(t, err, "Expected to receive an error")
+
+		resp := uitest.JSONUIFromBytes(t, []byte(out))
+
+		// if no changes are made, then wait for the existing version to reconcile (as it failed previously)
+		expectedOutput := []map[string]string{{
+			"kind":            "Job",
+			"name":            "my-job-ver-1",
+			"namespace":       env.Namespace,
+			"op":              "",
+			"op_strategy":     "",
+			"reconcile_info":  "Failed with reason\nBackoffLimitExceeded: Job has\nreached the specified backoff limit",
+			"reconcile_state": "fail",
+			"wait_to":         "reconcile",
+		}}
+
+		validateChanges(t, resp.Tables, expectedOutput, "Op:      0 create, 0 delete, 0 update, 1 noop, 0 exists",
+			"Wait to: 1 reconcile, 0 delete, 0 noop", out)
+	})
+
+	logger.Section("deploy new version with a change", func() {
+		out, _ := kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name, "--json"},
+			RunOpts{IntoNs: true, StdinReader: strings.NewReader(successYaml)})
+
+		resp := uitest.JSONUIFromBytes(t, []byte(out))
+
+		// if a change is made, i.e a new version is created, then kapp should ignore previous failed versions
+		expectedOutput := []map[string]string{{
+			"kind":            "Job",
+			"name":            "my-job-ver-2",
+			"namespace":       env.Namespace,
+			"op":              "create",
+			"op_strategy":     "",
+			"reconcile_info":  "",
+			"reconcile_state": "",
+			"wait_to":         "reconcile",
+		}}
+
+		validateChanges(t, resp.Tables, expectedOutput, "Op:      1 create, 0 delete, 0 update, 0 noop, 0 exists",
+			"Wait to: 1 reconcile, 0 delete, 0 noop", out)
+	})
+}

